### ISSUE TITLE
It seems that θ is not updated after initialization, and the values of the optimizer are not passed to θ. Details in body.

### ISSUE THREAD

**Issue by nyanyaxia** [adriengoldszal/MAP_578_ACIRL#3]
![image](https://github.com/user-attachments/assets/386b2205-c248-4958-9cb1-cfec6aa06b59)
Hi,
In maxnet.py, θ is not updated, which causes the code to break out of the loop after the second iteration. I don't know if it's my issue or if the code itself is designed this way.
Regards.

**Comment by adriengoldszal** [comment]
Hello, you are in fact correct. The optimizer has normally just been fixed. The optimization process is however pretty unstable. A low learning rate seems to do the trick for now but we might look into some form of learning rate decay.

**Comment by nyanyaxia** [comment]
Thank you!
Furthermore, I noticed that the indices of the state transition matrix and the feature matrix seem to be mismatched. The index calculation formula for the state transition matrix is y * size + x, while the index calculation formula for the feature matrix is x * size + y. The discrepancy will cause issues during inverse reinforcement learning.
I guess you should use x*size+y.

**Comment by adriengoldszal** [comment]
Thank you for finding that ! I uniformized the call to the function, it should be good now. Please do let us know if you have any other feedback !

**Comment by nyanyaxia** [comment]
I appreciate your acceptance of my feedback, but I suggest using x*size+y, as you also convert states to indices in maxent_wrapper.py, which is inconsistent with the code in gridworld.py.
![20241220111252](https://github.com/user-attachments/assets/859fb3c9-dd4d-4a84-a904-73c5315e0008)

**Comment by adriengoldszal** [comment]
Yes, you are right. I changed it so it is consistent now normally. Thank you for pointing it out !

**Comment by nyanyaxia** [comment]
Hello.
You set the learning rate too small, which results in insufficient update magnitude. The program mistakenly assumes that theta has converged and exits the loop directly. I suggest increasing the learning rate and reducing the threshold.
Looking forward to your response.
![QQ截图20241226164423](https://github.com/user-attachments/assets/dceaa56b-b7ec-4544-aaac-7626227ff3e9)

**Comment by nyanyaxia** [comment]
> Hello. You set the learning rate too small, which results in insufficient update magnitude. The program mistakenly assumes that theta has converged and exits the loop directly. I suggest increasing the learning rate and reducing the threshold. Looking forward to your response. ![QQ截图20241226164423](https://private-user-images.githubusercontent.com/67639368/398684460-dceaa56b-b7ec-4544-aaac-7626227ff3e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUyMDMxMjUsIm5iZiI6MTczNTIwMjgyNSwicGF0aCI6Ii82NzYzOTM2OC8zOTg2ODQ0NjAtZGNlYWE1NmItYjdlYy00NTQ0LWFhYWMtNzYyNjIyN2ZmM2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI2VDA4NDcwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwNDdlMzBjNGZkOTc0MjE1YzE1ZDFiNGRiZWI2YmNkNmU2NjM2YzQ1NmRhNDMxNGVkODMyNmQ4YjhhMWQ0MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pnQoXu841bskMTNvjbfwH-rB3vQhs2MXMZk1VCR4znw)

This can also explain why the training is not stable, as the program exits the loop before convergence.

**Comment by adriengoldszal** [comment]
Hello, yes, this seems to be the case sometimes, you are right, however, the learning has a tendency to diverge pretty quickly if the learning rate is higher, as you can see here (lr = 0.01, eps 1e-4)
<img width="263" alt="divergence" src="https://github.com/user-attachments/assets/6e5120c9-cc8a-4a7e-aa7a-96071557b1b5" />
I'll look into it after the christmas break in a few days, and see what can be done, tmaybe a decaying learning rate could help for example.

**Comment by nyanyaxia** [comment]
I see. Anyway, Merry Christmas!

**Comment by adriengoldszal** [comment]
Hello, I changed the optimizer on the test_optim branch which manages to converge, and changed the terminal states to the correct one, but doing so makes the convergence hard again. The gradients are also very high, which would mean the convergence isn't properly working, but I can't seem to fix that for now
